### PR TITLE
EMP Grenade actually sabotages power

### PR DESCRIPTION
**Pull request by WarMechanic**
## About the PR

EMP has been changed to target APC-powered-devices (most electrical devices) as well as batteries to disable them. This means EMP can interfere with autolathes, airlocks, atmos devices, substations and SMES. The power draw of a single EMP grenade now cuts out a substation, and the disabling effect prevents further recharge until it subsides.

EMP duration now also stacks, which creates a novel way to quietly black out the station by attacking engineering SMES with 3 EMP grenades (6tc EMP bundle) to black out the station for 3 minutes.

Edit, here's a detailed changelog of the PR,
Functionality:
- EMP disable has been generalised to kill and prevent further function of every device/battery by interrupting recharge
- As a result of the above, some hard coded interactions have been culled
- EMP disable duration now stacks with multiple EMP blasts
- EMP is now capable of draining from gravity generators
- The Charger system has been slightly reworked to facilitate communication between batteries and chargers

Results:
- EMP grenade can disable basically every powered machine, most notably doors
- EMP grenade has had its power drain upped to 2.7MW, which is slightly more than a substation and 1/3 a SMES
- EMP grenade can now instantly kill substations
- EMP grenade can now instantly kill gravity generators
- 3 EMP grenades (6tc) can be used to kill SMES and disable recharge for 3 minutes with no evidence on the power monitor.

## Why / Balance

EMP at 2tc has a relatively low value-proposition when compared to C4 which is also 2tc. While EMP can probably black out one (or two if you're lucky) APCs and can be used as a defensive option against Stun/Lasers. C4 can be used to cut wires, substations, SMES, generators, doors, reinforced walls, people and the list probably continues.

New EMP can be used to soft-bomb station power in an explosion that isn't globally alarming (salv boom). Targeting the captain's office directly may let you crowbar in and steal the locker but it leaves ephemeral evidence in the form of everything electrical shimmering blue. Opting to bomb substations blacks out a wider area, providing several degrees of separation from your target. That is to say, new EMP grenade favours map knowledge and rewards better stealth.

## Technical details

- `C.S/.../EmpSystem.cs` uses TryComp to turn on/off charging for `C.S/Power/Components/PowerNetworkBatteryComponent`
- `C.S/Power/EntitySystems/PowerReceiverSystem.cs` listens to `EmpPulseEvent` to turn off. Requests to turn back on are additionally intercepted by `EmpSystem.cs` and cancelled.
- `C.S/.../GravityGeneratorSystem.cs` listens to `EmpPulseEvent` and converts energy consumption to a normalised charge
- `C.S/Power/EntitySystems/ApcSystem.cs` no longer toggles its breaker, but still listens to `EmpPulseEvent` for updating visuals.
- `C.S/Power/EntitySystems/ChargerSystem.cs` was refactored to add a `ChargingComponent` flag to power cells instead of `ActiveCharger` on itself. Battery and Charger communicate through this flag. Listens to `EmpPulseEvent` for updating its state machine. New `ChargerUpdateStatusEvent` allows batteries to update the charger's status.
- `C.S/Power/EntitySystems/BatterySystem.cs` can now be disabled, and checks for disabling before updating its charge. Raises `ChargerUpdateStatusEvent` when hearing `EmpDisabledRemoved` to tell its charger to start charging again.
- `C.S/Power/PowerWireAction.cs` checks for `EmpDisabledComponent` before turning power back on.
- `C.S/SurveillanceCamera/Systems/SurveillanceCameraSystem.cs` and `C.S/VendingMachines/VendingMachineSystem.cs` had redundant `EmpPulseEvent` listeners culled.
- `Resources/Prototypes/Entities/Objects/Weapons/Throwable/grenades.yml` buffed EMP grenade.

## Media

https://www.youtube.com/embed/rSVph6OIg1s?si=8o4bx9Vx16B6usuu

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- tweak: EMP Grenades can now disable basically any electrical device, and stack in disable duration.

### CODE REVIEW

**Comment by WarMechanic** [comment]
As a quick note: due to `EmpSystem` now disabling machines there are a few obsolete specific interactions in `EmpSystem` which include: making vendors eject items randomly (???) and turning off cameras (they are turned off anyway). Should these be removed?

**Comment by Dutch-VanDerLinde** [comment]
changelog is broken

**Comment by WarMechanic** [comment]
im currently investigating an issue in that apc visuals don't update until after you inspect it

**Comment by WarMechanic** [comment]
should emp disable energy sword and derivatives for consistency? i think i can also do that here

**Comment by WarMechanic** [comment]
small refactor of how the power cell charger back-end works, it should be functionally the same in-game but the battery and charger can now communicate properly which was used to update the charger's state

**Comment by WarMechanic** [comment]
emp now correctly calculates power drain from gravity generators, and emp grenades are capable of instantly silencing gravity generators which is something that c4 can't do

might make for some solid nukie tactics in the future

**Comment by WarMechanic** [comment]
i could fix this https://github.com/space-wizards/space-station-14/issues/29116 now but would that be stretching the bounds of this PR?

**Comment by WarMechanic** [comment]
alright after some strenuous testing i think emp interactions with everything should be good to go

**Comment by WarMechanic** [comment]
i wanted to make emp affect energy sword too but that would require very specific overrides on the generic toggle system which is shit, it would be easier to make energy sword take power which i think would be pretty cool

**Comment by WarMechanic** [comment]
also whoever is reviewing this:
i do use the obsolete `someComponent.Owner` in my code simply because i wasn't sure how to handle it another way, if you want to save some time reviewing a suggested change will get me started on the rest

**Comment by K-Dynamic** [comment]
> i wanted to make emp affect energy sword too but that would require very specific overrides on the generic toggle system which is shit, it would be easier to make energy sword take power which i think would be pretty cool

I don't recommend this considering syndies/nukies use EMPs and it would be annoying, especially as eswords are a bit weak currently

**Comment by WarMechanic** [comment]
i'm considering expanding nanotrasen use of emp tech, perhaps through research. one thing im considering is an emp slug that can be utilised for door breaching, anti-silicon or in this case a countermeasure to syndicate energy weapons

i don't think esword emp vulnerability is a door that necessarily needs to be closed, but i can always revisit it downstream if its unwanted

File: Content.Server/Power/EntitySystems/ChargerSystem.cs
Comment (patrickkelly0) [inline Content.Server/Power/EntitySystems/ChargerSystem.cs:1]: this is really flipping cool

Comment (patrickkelly0) [reply Content.Server/Power/EntitySystems/ChargerSystem.cs:1]: +rep

**Comment by Boaz1111** [comment]
> i'm considering expanding nanotrasen use of emp tech, perhaps through research. one thing im considering is an emp slug that can be utilised for door breaching, anti-silicon or in this case a countermeasure to syndicate energy weapons
> 
> 
> 
> i don't think esword emp vulnerability is a door that necessarily needs to be closed, but i can always revisit it downstream if its unwanted

As far as NT EMP tech, ion rifle should coming in relatively soon

File: Content.Server/Gravity/GravityGeneratorSystem.cs [hunk 1]
```diff
@@ -286,5 +289,28 @@ private void MakeOn(Entity<GravityGeneratorComponent> ent, AppearanceComponent?
...
         }
+
+        private void OnEmpPulse(EntityUid uid, GravityGeneratorComponent component, EmpPulseEvent args)
+        {
+            /// i really don't think that the gravity generator should use normalised 0-1 charge
+            /// as opposed to watts charge that every other battery uses
+
+            ApcPowerReceiverComponent? powerReceiver = null;
+            if (!Resolve(uid, ref powerReceiver, false))
+                return;
```
Comment (metalgearsloth) [inline Content.Server/Gravity/GravityGeneratorSystem.cs:300]: Just trycomp this is pointless

File: Content.Server/Emp/EmpSystem.cs [hunk 2]
```diff
@@ -77,7 +75,19 @@ public void DoEmpEffects(EntityUid uid, float energyConsumption, float duration)
         if (ev.Disabled)
         {
             var disabled = EnsureComp<EmpDisabledComponent>(uid);
-            disabled.DisabledUntil = Timing.CurTime + TimeSpan.FromSeconds(duration);
+            // couldnt use null-coalescing operator here sadge
+            if (disabled.DisabledUntil == TimeSpan.Zero)
```
Comment (metalgearsloth) [inline Content.Server/Emp/EmpSystem.cs:79]: Just make it null

File: Content.Server/Power/EntitySystems/BatterySystem.cs [hunk 3]
```diff
@@ -100,9 +102,21 @@ private void CalculateBatteryPrice(EntityUid uid, BatteryComponent component, re
...
+        // if a disabled battery is put into a recharged,
+        // allow the recharger to start recharging again after the disable ends
+        private void OnEmpDisabledRemoved(EntityUid uid, BatteryComponent component, ref EmpDisabledRemoved args)
+        {
+            if (!TryComp<ChargingComponent>(uid, out var charging))
+                return;
+
+            var ev = new ChargerUpdateStatusEvent(); 
+            RaiseLocalEvent<ChargerUpdateStatusEvent>(charging.ChargerUid, ref ev);
+        }
```
Comment (metalgearsloth) [inline Content.Server/Power/EntitySystems/BatterySystem.cs:118]: Generic is pointless here afaik.

File: Content.Server/Power/EntitySystems/BatterySystem.cs [hunk 4]
```diff
@@ -157,6 +171,18 @@ public bool TryUseCharge(EntityUid uid, float value, BatteryComponent? battery =
...
         }
 
+        /// <summary>
+        ///     Like SetCharge, but checks for conditions like EmpDisabled before executing
+        /// </summary>
+        public bool TrySetCharge(EntityUid uid, float value, BatteryComponent? battery = null)
+        {
+            if (!Resolve(uid, ref battery, false) || TryComp<EmpDisabledComponent>(uid, out var emp))
+                return false;
+
```
Comment (metalgearsloth) [inline Content.Server/Power/EntitySystems/BatterySystem.cs:181]: Use the other version of trycomp here.

File: Content.Server/Power/EntitySystems/BatterySystem.cs [hunk 5]
```diff
@@ -165,6 +191,10 @@ public bool IsFull(EntityUid uid, BatteryComponent? battery = null)
             if (!Resolve(uid, ref battery))
                 return false;
 
+            // If the battery is full, remove its charging component.
+            if (TryComp<ChargingComponent>(uid, out _))
+                RemComp<ChargingComponent>(uid);
+
```
Comment (metalgearsloth) [inline Content.Server/Power/EntitySystems/BatterySystem.cs:197]: RemComp doesn't exception if it doesn't exist, just remcomp.

File: Content.Server/Power/EntitySystems/ChargerSystem.cs [hunk 6]
```diff
@@ -46,21 +49,58 @@ private void OnChargerExamine(EntityUid uid, ChargerComponent component, Examine
...
+        }
+
+        // ...so the status always updates (for insertin a power cell)
+        UpdateStatus(uid, component);
+    }
+
+    private void StopChargingBattery(EntityUid uid, ChargerComponent component, EntityUid target)
+    {
+        if (HasComp<ChargingComponent>(target))
+            RemComp<ChargingComponent>(target);
```
Comment (metalgearsloth) [inline Content.Server/Power/EntitySystems/ChargerSystem.cs:80]: Same here.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by slarticodefast** [comment]
Are you coming back to this?

**Comment by slarticodefast** [comment]
If you plan to continue working on this, please write a comment here and I will reopen the PR for you.

**Pull request closed without merging**